aws/resource_aws_backup_vault.go

```diff
@@ -0,0 +1,120 @@
+package aws
+
+import (
+	"regexp"
+
+	"github.com/hashicorp/terraform/helper/validation"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/backup"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsBackupVault() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsBackupVaultCreate,
+		Read:   resourceAwsBackupVaultRead,
+		Update: resourceAwsBackupVaultUpdate,
+		Delete: resourceAwsBackupVaultDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile(`[a-z0-9\-]+`), "must contain alphanumeric characters or underscores"),
+			},
+			"tags": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"kms_key_arn": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateArn,
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"recovery_points": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsBackupVaultCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).backupconn
+
+	input := &backup.CreateBackupVaultInput{
+		BackupVaultName: aws.String(d.Get("name").(string)),
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
+		input.BackupVaultTags = remapTags(v.(map[string]interface{}))
+	}
+
+	if v, ok := d.GetOk("kms_key_arn"); ok {
+		input.EncryptionKeyArn = aws.String(v.(string))
+	}
+
+	_, err := conn.CreateBackupVault(input)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(d.Get("name").(string))
+
+	return resourceAwsBackupVaultRead(d, meta)
+}
+
+func resourceAwsBackupVaultRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).backupconn
+
+	input := &backup.DescribeBackupVaultInput{
+		BackupVaultName: aws.String(d.Id()),
+	}
+
+	resp, err := conn.DescribeBackupVault(input)
+	if err != nil {
+		return err
+	}
+
+	d.Set("arn", resp.BackupVaultArn)
+	d.Set("recovery_points", resp.NumberOfRecoveryPoints)
+
+	return nil
+}
+
+func resourceAwsBackupVaultUpdate(d *schema.ResourceData, meta interface{}) error {
+	return resourceAwsBackupVaultRead(d, meta)
+}
+
+func resourceAwsBackupVaultDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).backupconn
+
+	input := &backup.DeleteBackupVaultInput{
+		BackupVaultName: aws.String(d.Get("name").(string)),
+	}
+
+	_, err := conn.DeleteBackupVault(input)
+	if err != nil {
+		return err
```

Comment:
We should provide error context for operators and code maintainers, e.g.

```suggestion
		return fmt.Errorf("error deleting Backup Vault (%s): %s", d.Id(), err)
```